conn.go

```diff
@@ -35,3 +40,110 @@ func (c *conn) RemotePeer() peer.ID {
 func (c *conn) RemotePublicKey() ci.PubKey {
 	return c.remotePubKey
 }
+
+const (
+	recordTypeHandshake byte = 22
+	versionTLS13             = 0x0304
+	maxCiphertextTLS13       = 16384 + 256 // maximum ciphertext length in TLS 1.3
+	maxHandshake             = 65536       // maximum handshake we support (protocol max is 16 MB)
+)
+
+var errSimultaneousConnect = errors.New("detected TCP simultaneous connect")
+
+type teeConn struct {
+	net.Conn
+	buf *bytes.Buffer
+}
+
+func newTeeConn(c net.Conn, buf *bytes.Buffer) net.Conn {
+	return &teeConn{Conn: c, buf: buf}
+}
+
+func (c *teeConn) Read(b []byte) (int, error) {
+	n, err := c.Conn.Read(b)
+	c.buf.Write(b[:n])
+	return n, err
+}
+
+type wrappedConn struct {
+	net.Conn
+
+	hasReadFirstMessage bool
+	raw                 bytes.Buffer // contains a copy of every byte of the first handshake message we read from the wire
+
+	hand bytes.Buffer // used to store the first handshake message until we've completely read it
+
+}
+
+func newWrappedConn(c net.Conn) net.Conn {
+	wc := &wrappedConn{}
+	wc.Conn = newTeeConn(c, &wc.raw)
+	return wc
+}
+
+func (c *wrappedConn) Read(b []byte) (int, error) {
+	if c.hasReadFirstMessage {
+		return c.Conn.Read(b)
```

Comment:
won't this keep reading through the `teeConn`?

Reply:
(and more importantly keep adding to that buffer)

Reply:
You're right! Thanks for catching this!
I updated the logic such that we switch the `wrappedConn.Conn` over to the un-tee-d connection once we're done reading the first handshake message.